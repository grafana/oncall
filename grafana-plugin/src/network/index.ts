import { SpanStatusCode } from '@opentelemetry/api';
import { SemanticAttributes } from '@opentelemetry/semantic-conventions';
import axios from 'axios';
import qs from 'query-string';

import FaroHelper from 'utils/faro';

export const API_HOST = `${window.location.protocol}//${window.location.host}/`;
export const API_PROXY_PREFIX = 'api/plugin-proxy/grafana-oncall-app';
export const API_PATH_PREFIX = '/api/internal/v1';

const instance = axios.create();

instance.interceptors.request.use(function (config) {
  // Do something before request is sent
  config.paramsSerializer = (params) => {
    return qs.stringify(params, { arrayFormat: 'none' });
  };

  config.validateStatus = (status) => {
    return status >= 200 && status < 300; // default
  };

  return {
    ...config,
  };
});

interface RequestConfig {
  method?: 'GET' | 'POST' | 'PUT' | 'DELETE' | 'OPTIONS';
  params?: any;
  data?: any;
  withCredentials?: boolean;
  validateStatus?: (status: number) => boolean;
}

export const isNetworkError = axios.isAxiosError;

export const makeRequest = async <RT = any>(path: string, config: RequestConfig) => {
  const { method = 'GET', params, data, validateStatus } = config;

  const url = `${API_PROXY_PREFIX}${API_PATH_PREFIX}${path}`;
  const otel = FaroHelper.faro?.api?.getOTEL();

  if (FaroHelper.faro && otel) {
    const tracer = otel.trace.getTracer('default');
    let span = otel.trace.getActiveSpan();

    if (!span) {
      span = tracer.startSpan('http-request');
      span.setAttribute('page_url', document.URL.split('//')[1]);
      span.setAttribute(SemanticAttributes.HTTP_URL, url);
      span.setAttribute(SemanticAttributes.HTTP_METHOD, method);
    }

    return new Promise<RT>((resolve, reject) => {
      otel.context.with(otel.trace.setSpan(otel.context.active(), span), async () => {
        FaroHelper.faro.api.pushEvent('Sending request', { url });

        instance({
          method,
          url,
          params,
          data,
          validateStatus,
        })
          .then((response) => {
            FaroHelper.faro.api.pushEvent('Request completed', { url });
            span.end();
            resolve(response.data as RT);
          })
          .catch((ex) => {
            FaroHelper.faro.api.pushEvent('Request failed', { url });
            FaroHelper.faro.api.pushError(ex);
            span.setStatus({ code: SpanStatusCode.ERROR });
            span.end();
            reject(ex);
          });
      });
    });
  }

  return instance({
    method,
    url,
    params,
    data,
    validateStatus,
  })
    .then((response) => {
      FaroHelper.faro?.api.pushEvent('Request completed', { url });
      return response.data as RT;
    })
    .catch((ex) => {
      FaroHelper.faro?.api.pushEvent('Request failed', { url });
      FaroHelper.faro?.api.pushError(ex);
      return Promise.reject(ex);
    });
};
