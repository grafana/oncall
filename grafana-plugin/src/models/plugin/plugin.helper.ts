import { makeRequest } from 'network/network';

export class PluginHelper {
  static async install() {
    return makeRequest(`/plugin/install`, {
      method: 'POST',
    });
  }
}
