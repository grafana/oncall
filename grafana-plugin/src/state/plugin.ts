import { getBackendSrv } from '@grafana/runtime';

import { makeRequest } from 'network';

export async function createGrafanaToken() {
  const keys = await getBackendSrv().get('/api/auth/keys');
  const existingKey = keys.find((key: { id: number; name: string; role: string }) => key.name === 'OnCall');

  if (existingKey) {
    await getBackendSrv().delete(`/api/auth/keys/${existingKey.id}`);
  }

  return await getBackendSrv().post('/api/auth/keys', {
    name: 'OnCall',
    role: 'Admin',
    secondsToLive: null,
  });
}

export async function updateGrafanaToken(key: string) {
  await getBackendSrv().post(`/api/plugins/grafana-oncall-app/settings`, {
    enabled: true,
    pinned: true,
    secureJsonData: {
      grafanaToken: key,
    },
  });
}

export async function startPluginSync() {
  return await makeRequest('/plugin/sync', { method: 'POST' });
}

export const SYNC_STATUS_RETRY_LIMIT = 10;

export const syncStatusDelay = (retryCount) => new Promise((resolve) => setTimeout(resolve, 10 * 2 ** retryCount));

export async function getPluginSyncStatus() {
  return await makeRequest(`/plugin/sync`, { method: 'GET' });
}

export async function installPlugin() {
  const grafanaToken = await createGrafanaToken();
  await updateGrafanaToken(grafanaToken.key);
  return await makeRequest('/plugin/install', { method: 'POST' });
}
