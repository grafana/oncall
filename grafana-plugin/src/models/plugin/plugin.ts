import { isEqual } from 'lodash-es';
import { makeAutoObservable, runInAction } from 'mobx';
import { OnCallPluginMetaJSONData } from 'types';

import { ActionKey } from 'models/loader/action-keys';
import { GrafanaApiClient } from 'network/grafana-api/http-client';
import { makeRequest } from 'network/network';
import { PluginConnection, StatusResponse } from 'network/oncall-api/api.types';
import { RootBaseStore } from 'state/rootBaseStore/RootBaseStore';
import { waitInMs } from 'utils/async';
import { AutoLoadingState } from 'utils/decorators';

import { PluginHelper } from './plugin.helper';

/* 
High-level OnCall initialization process:
On OSS:
  - On OnCall page / OnCall extension mount POST /status is called and it has pluginConfiguration object with different flags. 
    If all of them have `ok: true` , we consider plugin to be successfully configured and application loading is being continued. 
    Otherwise, we show error page with the option to go to plugin config (for Admin user) or to contact administrator (for nonAdmin user)
  - On plugin config page frontend sends another POST /status. If every flag has `ok: true`, it shows that plugin is connected. 
    Otherwise, it shows more detailed information of what is misconfigured / missing. User can update onCallApiUrl and try to reconnect plugin.
      - If Grafana version >= 10.3 AND externalServiceAccount feature flag is `true`, then grafana token is autoprovisioned and there is no need to create it
      - Otherwise, user is given the option to manually create service account as Admin and then reconnect the plugin
On Cloud:
  - On OnCall page / OnCall extension mount POST /status is called. If plugin is configured correctly, application loads as usual.
    If it's not, we show error page with the button to contact support
  - On plugin config page we show info if plugin is connected. If it's not we show detailed information of the errors and the button to contact support
*/

export class PluginStore {
  rootStore: RootBaseStore;
  connectionStatus?: PluginConnection;
  apiUrlFromStatus?: string;
  isPluginConnected = false;
  appliedOnCallApiUrl = '';

  constructor(rootStore: RootBaseStore) {
    makeAutoObservable(this, undefined, { autoBind: true });
    this.rootStore = rootStore;
  }

  private resetConnectionStatus() {
    this.connectionStatus = undefined;
    this.isPluginConnected = false;
  }

  async refreshAppliedOnCallApiUrl() {
    const { jsonData } = await GrafanaApiClient.getGrafanaPluginSettings();
    runInAction(() => {
      this.appliedOnCallApiUrl = jsonData.onCallApiUrl;
    });
  }

  @AutoLoadingState(ActionKey.PLUGIN_VERIFY_CONNECTION)
  async verifyPluginConnection() {
    const { pluginConnection, api_url } = await makeRequest<StatusResponse>(`/plugin/status`, {});
    runInAction(() => {
      this.connectionStatus = pluginConnection;
      this.apiUrlFromStatus = api_url;
      this.isPluginConnected = Object.keys(pluginConnection).every(
        (key) => pluginConnection[key as keyof PluginConnection]?.ok
      );
    });
  }

  @AutoLoadingState(ActionKey.PLUGIN_UPDATE_SETTINGS_AND_REINITIALIZE)
  async updatePluginSettingsAndReinitializePlugin({
    currentJsonData,
    newJsonData,
  }: {
    currentJsonData: OnCallPluginMetaJSONData;
    newJsonData: Partial<OnCallPluginMetaJSONData>;
  }) {
    this.resetConnectionStatus();
    const saveJsonDataCandidate = { ...currentJsonData, ...newJsonData };
    if (!isEqual(currentJsonData, saveJsonDataCandidate) || !this.connectionStatus?.oncall_api_url?.ok) {
      await GrafanaApiClient.updateGrafanaPluginSettings({ jsonData: saveJsonDataCandidate });
      await waitInMs(1000); // It's required for backend proxy to pick up new settings
    }
    try {
      await PluginHelper.install();
    } finally {
      await this.verifyPluginConnection();
    }
  }

  @AutoLoadingState(ActionKey.PLUGIN_RECREATE_SERVICE_ACCOUNT)
  async recreateServiceAccountAndRecheckPluginStatus() {
    await GrafanaApiClient.recreateGrafanaTokenAndSaveInPluginSettings();
    await this.verifyPluginConnection();
  }

  async enablePlugin() {
    await GrafanaApiClient.updateGrafanaPluginSettings({}, true);
    location.reload();
  }
}
