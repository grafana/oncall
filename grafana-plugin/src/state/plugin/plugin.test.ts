import { makeRequest as makeRequestOriginal, isNetworkError as isNetworkErrorOriginal } from 'network/network';

import { PluginState, InstallationVerb, UpdateGrafanaPluginSettingsProps } from './plugin';

const makeRequest = makeRequestOriginal as jest.Mock<ReturnType<typeof makeRequestOriginal>>;
const isNetworkError = isNetworkErrorOriginal as unknown as jest.Mock<ReturnType<typeof isNetworkErrorOriginal>>;

jest.mock('network/network');

afterEach(() => {
  jest.resetAllMocks();
});

const ONCALL_BASE_URL = '/plugin';
const GRAFANA_PLUGIN_SETTINGS_URL = '/api/plugins/grafana-oncall-app/settings';

const generateMockNetworkError = (status: number, data = {}) => ({ response: { status, ...data } });

describe('PluginState.generateOnCallApiUrlConfiguredThroughEnvVarMsg', () => {
  test.each([true, false])(
    'it returns the proper error message - configured through env var: %s',
    (configuredThroughEnvVar) => {
      expect(PluginState.generateOnCallApiUrlConfiguredThroughEnvVarMsg(configuredThroughEnvVar)).toMatchSnapshot();
    }
  );
});

describe('PluginState.generateInvalidOnCallApiURLErrorMsg', () => {
  test.each([true, false])(
    'it returns the proper error message - configured through env var: %s',
    (configuredThroughEnvVar) => {
      expect(
        PluginState.generateInvalidOnCallApiURLErrorMsg('http://hello.com', configuredThroughEnvVar)
      ).toMatchSnapshot();
    }
  );
});

describe('PluginState.generateUnknownErrorMsg', () => {
  test.each([
    [true, 'install'],
    [true, 'sync'],
    [false, 'install'],
    [false, 'sync'],
  ])(
    'it returns the proper error message - configured through env var: %s',
    (configuredThroughEnvVar, verb: InstallationVerb) => {
      expect(PluginState.generateUnknownErrorMsg('http://hello.com', verb, configuredThroughEnvVar)).toMatchSnapshot();
    }
  );
});

describe('PluginState.getHumanReadableErrorFromOnCallError', () => {
  beforeEach(() => {
    console.warn = () => {};
  });

  test.each([502, 409])('it handles a non-400 network error properly - status code: %s', (status) => {
    isNetworkError.mockReturnValueOnce(true);

    expect(
      PluginState.getHumanReadableErrorFromOnCallError(
        generateMockNetworkError(status),
        'http://hello.com',
        'install',
        true
      )
    ).toMatchSnapshot();
  });

  test.each([true, false])(
    'it handles a 400 network error properly - has custom error message: %s',
    (hasCustomErrorMessage) => {
      isNetworkError.mockReturnValueOnce(true);

      const networkError = generateMockNetworkError(400) as any;
      if (hasCustomErrorMessage) {
        networkError.response.data = { error: 'ohhhh nooo an error' };
      }
      expect(
        PluginState.getHumanReadableErrorFromOnCallError(networkError, 'http://hello.com', 'install', true)
      ).toMatchSnapshot();
    }
  );

  test('it handles an unknown error properly', () => {
    isNetworkError.mockReturnValueOnce(false);

    expect(
      PluginState.getHumanReadableErrorFromOnCallError(new Error('asdfasdf'), 'http://hello.com', 'install', true)
    ).toMatchSnapshot();
  });
});

describe('PluginState.getHumanReadableErrorFromGrafanaProvisioningError', () => {
  beforeEach(() => {
    console.warn = () => {};
  });

  test.each([true, false])('it handles an error properly - network error: %s', (networkError) => {
    const onCallApiUrl = 'http://hello.com';
    const installationVerb = 'install';
    const onCallApiUrlIsConfiguredThroughEnvVar = true;
    const error = networkError ? generateMockNetworkError(400) : new Error('oh noooo');

    const mockGenerateInvalidOnCallApiURLErrorMsgResult = 'asdadslkjfkjlsd';
    const mockGenerateUnknownErrorMsgResult = 'asdadslkjfkjlsd';

    isNetworkError.mockReturnValueOnce(networkError);

    PluginState.generateInvalidOnCallApiURLErrorMsg = jest
      .fn()
      .mockReturnValueOnce(mockGenerateInvalidOnCallApiURLErrorMsgResult);
    PluginState.generateUnknownErrorMsg = jest.fn().mockReturnValueOnce(mockGenerateUnknownErrorMsgResult);

    const expectedErrorMsg = networkError
      ? mockGenerateInvalidOnCallApiURLErrorMsgResult
      : mockGenerateUnknownErrorMsgResult;

    expect(
      PluginState.getHumanReadableErrorFromGrafanaProvisioningError(
        error,
        onCallApiUrl,
        installationVerb,
        onCallApiUrlIsConfiguredThroughEnvVar
      )
    ).toEqual(expectedErrorMsg);

    if (networkError) {
      expect(PluginState.generateInvalidOnCallApiURLErrorMsg).toHaveBeenCalledTimes(1);
      expect(PluginState.generateInvalidOnCallApiURLErrorMsg).toHaveBeenCalledWith(
        onCallApiUrl,
        onCallApiUrlIsConfiguredThroughEnvVar
      );
    } else {
      expect(PluginState.generateUnknownErrorMsg).toHaveBeenCalledTimes(1);
      expect(PluginState.generateUnknownErrorMsg).toHaveBeenCalledWith(
        onCallApiUrl,
        installationVerb,
        onCallApiUrlIsConfiguredThroughEnvVar
      );
    }
  });
});

describe('PluginState.getGrafanaPluginSettings', () => {
  test('it calls the proper method', async () => {
    PluginState.grafanaBackend.get = jest.fn();

    await PluginState.getGrafanaPluginSettings();

    expect(PluginState.grafanaBackend.get).toHaveBeenCalledTimes(1);
    expect(PluginState.grafanaBackend.get).toHaveBeenCalledWith(GRAFANA_PLUGIN_SETTINGS_URL);
  });
});

describe('PluginState.updateGrafanaPluginSettings', () => {
  test.each([true, false])('it calls the proper method - enabled: %s', async (enabled) => {
    const data: UpdateGrafanaPluginSettingsProps = {
      jsonData: {
        onCallApiUrl: 'asdfasdf',
      },
      secureJsonData: {
        grafanaToken: 'kjdfkfdjkffd',
      },
    };

    PluginState.grafanaBackend.post = jest.fn();

    await PluginState.updateGrafanaPluginSettings(data, enabled);

    expect(PluginState.grafanaBackend.post).toHaveBeenCalledTimes(1);
    expect(PluginState.grafanaBackend.post).toHaveBeenCalledWith(GRAFANA_PLUGIN_SETTINGS_URL, {
      ...data,
      enabled,
      pinned: true,
    });
  });
});

describe('PluginState.createGrafanaToken', () => {
  const cases = [
    [true, true, false],
    [true, false, false],
    [false, true, true],
    [false, true, false],
    [false, false, false],
  ];

  test.each(cases)(
    'it calls the proper methods - existing key: %s, existing sa: %s, existing token: %s',
    async (apiKeyExists, saExists, apiTokenExists) => {
      const baseUrl = PluginState.KEYS_BASE_URL;
      const serviceAccountBaseUrl = PluginState.SERVICE_ACCOUNTS_BASE_URL;
      const apiKeyId = 12345;
      const apiKeyName = PluginState.ONCALL_KEY_NAME;
      const apiKey = { name: apiKeyName, id: apiKeyId };
      const saId = 33333;
      const serviceAccount = { id: saId };

      PluginState.getGrafanaToken = jest.fn().mockReturnValueOnce(apiKeyExists ? apiKey : null);
      PluginState.grafanaBackend.delete = jest.fn();
      PluginState.grafanaBackend.post = jest.fn();

      PluginState.getServiceAccount = jest.fn().mockReturnValueOnce(saExists ? serviceAccount : null);
      PluginState.getOrCreateServiceAccount = jest.fn().mockReturnValueOnce(serviceAccount);
      PluginState.getTokenFromServiceAccount = jest.fn().mockReturnValueOnce(apiTokenExists ? apiKey : null);

      await PluginState.createGrafanaToken();

      expect(PluginState.getGrafanaToken).toHaveBeenCalledTimes(1);

      if (apiKeyExists) {
        expect(PluginState.grafanaBackend.delete).toHaveBeenCalledTimes(1);
        expect(PluginState.grafanaBackend.delete).toHaveBeenCalledWith(`${baseUrl}/${apiKey.id}`);
      } else if (apiTokenExists) {
        expect(PluginState.grafanaBackend.delete).toHaveBeenCalledTimes(1);
        expect(PluginState.grafanaBackend.delete).toHaveBeenCalledWith(
          `${serviceAccountBaseUrl}/${serviceAccount.id}/tokens/${apiKey.id}`
        );
      } else {
        expect(PluginState.grafanaBackend.delete).not.toHaveBeenCalled();
      }

      expect(PluginState.grafanaBackend.post).toHaveBeenCalledTimes(1);
      expect(PluginState.grafanaBackend.post).toHaveBeenCalledWith(
        `${serviceAccountBaseUrl}/${serviceAccount.id}/tokens`,
        {
          name: apiKeyName,
          role: 'Admin',
        }
      );
    }
  );
});

describe('PluginState.installPlugin', () => {
  it.each([true, false])('returns the proper response - self hosted: %s', async (selfHosted) => {
    // mocks
    const mockedResponse = 'asdfasdf';
    const grafanaToken = 'asdfasdf';
    const mockedCreateGrafanaTokenResponse = { key: grafanaToken };

    makeRequest.mockResolvedValueOnce(mockedResponse);
    PluginState.createGrafanaToken = jest.fn().mockResolvedValueOnce(mockedCreateGrafanaTokenResponse);
    PluginState.updateGrafanaPluginSettings = jest.fn();

    // test
    const response = await PluginState.installPlugin(selfHosted);

    // assertions
    expect(response).toEqual({
      grafanaToken,
      onCallAPIResponse: mockedResponse,
    });

    expect(PluginState.createGrafanaToken).toHaveBeenCalledTimes(1);
    expect(PluginState.createGrafanaToken).toHaveBeenCalledWith();

    expect(PluginState.updateGrafanaPluginSettings).toHaveBeenCalledTimes(1);
    expect(PluginState.updateGrafanaPluginSettings).toHaveBeenCalledWith({
      secureJsonData: {
        grafanaToken,
      },
    });

    expect(makeRequest).toHaveBeenCalledTimes(1);
    expect(makeRequest).toHaveBeenCalledWith(
      `${PluginState.ONCALL_BASE_URL}/${selfHosted ? 'self-hosted/' : ''}install`,
      {
        method: 'POST',
      }
    );
  });
});

describe('PluginState.selfHostedInstallPlugin', () => {
  test('it returns null if everything is successful', async () => {
    // mocks
    const onCallApiUrl = 'http://hello.com';
    const installPluginResponse = {
      grafanaToken: 'asldkaljkasdfjklfdasklj',
      onCallAPIResponse: {
        stackId: 5,
        orgId: 5,
        license: 'asdfasdf',
        onCallToken: 'asdfasdf',
      },
    };
    const {
      grafanaToken,
      onCallAPIResponse: { onCallToken: onCallApiToken, ...jsonData },
    } = installPluginResponse;

    PluginState.updateGrafanaPluginSettings = jest.fn();
    PluginState.installPlugin = jest.fn().mockResolvedValueOnce(installPluginResponse);

    // test
    const response = await PluginState.selfHostedInstallPlugin(onCallApiUrl, false);

    // assertions
    expect(response).toBeNull();

    expect(PluginState.updateGrafanaPluginSettings).toHaveBeenNthCalledWith(1, {
      jsonData: {
        onCallApiUrl,
      },
    });

    expect(PluginState.installPlugin).toHaveBeenCalledTimes(1);
    expect(PluginState.installPlugin).toHaveBeenCalledWith(true);

    expect(PluginState.updateGrafanaPluginSettings).toHaveBeenNthCalledWith(2, {
      jsonData: {
        ...jsonData,
        onCallApiUrl,
      },
      secureJsonData: {
        grafanaToken,
        onCallApiToken,
      },
    });
  });

  test('it returns an error msg if it cannot update the provisioning settings the first time around', async () => {
    // mocks
    const onCallApiUrl = 'http://hello.com';
    const mockedError = new Error('ohhh nooo');
    const mockedHumanReadableError = 'asdflkajsdflkajsdf';

    PluginState.updateGrafanaPluginSettings = jest.fn().mockRejectedValueOnce(mockedError);
    PluginState.getHumanReadableErrorFromGrafanaProvisioningError = jest
      .fn()
      .mockReturnValueOnce(mockedHumanReadableError);

    // test
    const response = await PluginState.selfHostedInstallPlugin(onCallApiUrl, false);

    // assertions
    expect(response).toEqual(mockedHumanReadableError);

    expect(PluginState.updateGrafanaPluginSettings).toHaveBeenCalledTimes(1);
    expect(PluginState.updateGrafanaPluginSettings).toHaveBeenCalledWith({
      jsonData: {
        onCallApiUrl,
      },
    });

    expect(PluginState.getHumanReadableErrorFromGrafanaProvisioningError).toHaveBeenCalledTimes(1);
    expect(PluginState.getHumanReadableErrorFromGrafanaProvisioningError).toHaveBeenCalledWith(
      mockedError,
      onCallApiUrl,
      'install',
      false
    );
  });

  test('it returns an error msg if it fails when installing the plugin,', async () => {
    // mocks
    const onCallApiUrl = 'http://hello.com';
    const mockedError = new Error('ohhh nooo');
    const mockedHumanReadableError = 'asdflkajsdflkajsdf';

    PluginState.updateGrafanaPluginSettings = jest.fn();
    PluginState.installPlugin = jest.fn().mockRejectedValueOnce(mockedError);
    PluginState.getHumanReadableErrorFromOnCallError = jest.fn().mockReturnValueOnce(mockedHumanReadableError);

    // test
    const response = await PluginState.selfHostedInstallPlugin(onCallApiUrl, false);

    // assertions
    expect(response).toEqual(mockedHumanReadableError);

    expect(PluginState.installPlugin).toHaveBeenCalledTimes(1);
    expect(PluginState.installPlugin).toHaveBeenCalledWith(true);

    expect(PluginState.getHumanReadableErrorFromOnCallError).toHaveBeenCalledTimes(1);
    expect(PluginState.getHumanReadableErrorFromOnCallError).toHaveBeenCalledWith(
      mockedError,
      onCallApiUrl,
      'install',
      false
    );
  });

  test('it returns an error msg if it cannot update the provisioning settings the second time around', async () => {
    // mocks
    const onCallApiUrl = 'http://hello.com';
    const mockedError = new Error('ohhh nooo');
    const mockedHumanReadableError = 'asdflkajsdflkajsdf';
    const installPluginResponse = {
      grafanaToken: 'asldkaljkasdfjklfdasklj',
      onCallAPIResponse: {
        stackId: 5,
        orgId: 5,
        license: 'asdfasdf',
        onCallToken: 'asdfasdf',
      },
    };
    const {
      grafanaToken,
      onCallAPIResponse: { onCallToken: onCallApiToken, ...jsonData },
    } = installPluginResponse;

    PluginState.updateGrafanaPluginSettings = jest.fn().mockResolvedValueOnce(null).mockRejectedValueOnce(mockedError);
    PluginState.installPlugin = jest.fn().mockResolvedValueOnce(installPluginResponse);
    PluginState.getHumanReadableErrorFromGrafanaProvisioningError = jest
      .fn()
      .mockReturnValueOnce(mockedHumanReadableError);

    // test
    const response = await PluginState.selfHostedInstallPlugin(onCallApiUrl, false);

    // assertions
    expect(response).toEqual(mockedHumanReadableError);

    expect(PluginState.updateGrafanaPluginSettings).toHaveBeenNthCalledWith(1, {
      jsonData: {
        onCallApiUrl,
      },
    });

    expect(PluginState.installPlugin).toHaveBeenCalledTimes(1);
    expect(PluginState.installPlugin).toHaveBeenCalledWith(true);

    expect(PluginState.updateGrafanaPluginSettings).toHaveBeenNthCalledWith(2, {
      jsonData: {
        ...jsonData,
        onCallApiUrl,
      },
      secureJsonData: {
        grafanaToken,
        onCallApiToken,
      },
    });

    expect(PluginState.getHumanReadableErrorFromGrafanaProvisioningError).toHaveBeenCalledTimes(1);
    expect(PluginState.getHumanReadableErrorFromGrafanaProvisioningError).toHaveBeenCalledWith(
      mockedError,
      onCallApiUrl,
      'install',
      false
    );
  });
});

describe('PluginState.updatePluginStatus', () => {
  test('it returns the API response', async () => {
    // mocks
    const mockedResp = { foo: 'bar' };
    const onCallApiUrl = 'http://hello.com';
    makeRequest.mockResolvedValueOnce(mockedResp);

    // test
    const response = await PluginState.updatePluginStatus(onCallApiUrl);

    // assertions
    expect(response).toEqual(mockedResp);

    expect(makeRequest).toHaveBeenCalledTimes(1);
    expect(makeRequest).toHaveBeenCalledWith(`${ONCALL_BASE_URL}/status`, { method: 'POST' });
  });

  test('it returns a human readable error in the event of an unsuccessful api call', async () => {
    // mocks
    const mockedError = new Error('hello');
    const mockedHumanReadableError = 'asdflkajsdflkajsdf';
    const onCallApiUrl = 'http://hello.com';
    makeRequest.mockRejectedValueOnce(mockedError);

    PluginState.getHumanReadableErrorFromOnCallError = jest.fn().mockReturnValueOnce(mockedHumanReadableError);

    // test
    const response = await PluginState.updatePluginStatus(onCallApiUrl);

    // assertions
    expect(response).toEqual(mockedHumanReadableError);

    expect(makeRequest).toHaveBeenCalledTimes(1);
    expect(makeRequest).toHaveBeenCalledWith(`${ONCALL_BASE_URL}/status`, { method: 'POST' });

    expect(PluginState.getHumanReadableErrorFromOnCallError).toHaveBeenCalledTimes(1);
    expect(PluginState.getHumanReadableErrorFromOnCallError).toHaveBeenCalledWith(
      mockedError,
      onCallApiUrl,
      'install',
      false
    );
  });
});

describe('PluginState.resetPlugin', () => {
  test('it calls grafanaBackend.post with the proper settings', async () => {
    // mocks
    const mockedResponse = 'asdfasdf';
    PluginState.updateGrafanaPluginSettings = jest.fn().mockResolvedValueOnce(mockedResponse);

    // test
    const response = await PluginState.resetPlugin();

    // assertions
    expect(response).toEqual(mockedResponse);
    expect(PluginState.updateGrafanaPluginSettings).toHaveBeenCalledTimes(1);
    expect(PluginState.updateGrafanaPluginSettings).toHaveBeenCalledWith(
      {
        jsonData: {
          stackId: null,
          orgId: null,
          onCallApiUrl: null,
          license: null,
        },
        secureJsonData: {
          grafanaToken: null,
          onCallApiToken: null,
        },
      },
      false
    );
  });
});
