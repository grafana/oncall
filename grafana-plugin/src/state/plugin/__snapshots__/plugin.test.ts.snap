// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PluginState.generateInvalidOnCallApiURLErrorMsg it returns the proper error message - configured through env var: false 1`] = `
"Could not communicate with OnCall API at http://hello.com.
Validate that the URL is correct, OnCall API is running, and that it is accessible from your Grafana instance."
`;

exports[`PluginState.generateInvalidOnCallApiURLErrorMsg it returns the proper error message - configured through env var: true 1`] = `
"Could not communicate with OnCall API at http://hello.com (NOTE: OnCall API URL is currently being taken from process.env of your UI).
Validate that the URL is correct, OnCall API is running, and that it is accessible from your Grafana instance."
`;

exports[`PluginState.generateOnCallApiUrlConfiguredThroughEnvVarMsg it returns the proper error message - configured through env var: false 1`] = `""`;

exports[`PluginState.generateOnCallApiUrlConfiguredThroughEnvVarMsg it returns the proper error message - configured through env var: true 1`] = `" (NOTE: OnCall API URL is currently being taken from process.env of your UI)"`;

exports[`PluginState.generateUnknownErrorMsg it returns the proper error message - configured through env var: false 1`] = `
"An unknown error occurred when trying to install the plugin. Verify OnCall API URL, http://hello.com, is correct?
Refresh your page and try again, or try removing your plugin configuration and reconfiguring."
`;

exports[`PluginState.generateUnknownErrorMsg it returns the proper error message - configured through env var: false 2`] = `
"An unknown error occurred when trying to sync the plugin. Verify OnCall API URL, http://hello.com, is correct?
Refresh your page and try again, or try removing your plugin configuration and reconfiguring."
`;

exports[`PluginState.generateUnknownErrorMsg it returns the proper error message - configured through env var: true 1`] = `
"An unknown error occurred when trying to install the plugin. Verify OnCall API URL, http://hello.com, is correct (NOTE: OnCall API URL is currently being taken from process.env of your UI)?
Refresh your page and try again, or try removing your plugin configuration and reconfiguring."
`;

exports[`PluginState.generateUnknownErrorMsg it returns the proper error message - configured through env var: true 2`] = `
"An unknown error occurred when trying to sync the plugin. Verify OnCall API URL, http://hello.com, is correct (NOTE: OnCall API URL is currently being taken from process.env of your UI)?
Refresh your page and try again, or try removing your plugin configuration and reconfiguring."
`;

exports[`PluginState.getHumanReadableErrorFromOnCallError it handles a 400 network error properly - has custom error message: false 1`] = `
"An unknown error occurred when trying to install the plugin. Verify OnCall API URL, http://hello.com, is correct (NOTE: OnCall API URL is currently being taken from process.env of your UI)?
Refresh your page and try again, or try removing your plugin configuration and reconfiguring."
`;

exports[`PluginState.getHumanReadableErrorFromOnCallError it handles a 400 network error properly - has custom error message: true 1`] = `"ohhhh nooo an error"`;

exports[`PluginState.getHumanReadableErrorFromOnCallError it handles a non-400 network error properly - status code: 409 1`] = `
"An unknown error occurred when trying to install the plugin. Verify OnCall API URL, http://hello.com, is correct (NOTE: OnCall API URL is currently being taken from process.env of your UI)?
Refresh your page and try again, or try removing your plugin configuration and reconfiguring."
`;

exports[`PluginState.getHumanReadableErrorFromOnCallError it handles a non-400 network error properly - status code: 502 1`] = `
"Could not communicate with OnCall API at http://hello.com (NOTE: OnCall API URL is currently being taken from process.env of your UI).
Validate that the URL is correct, OnCall API is running, and that it is accessible from your Grafana instance."
`;

exports[`PluginState.getHumanReadableErrorFromOnCallError it handles an unknown error properly 1`] = `
"An unknown error occurred when trying to install the plugin. Verify OnCall API URL, http://hello.com, is correct (NOTE: OnCall API URL is currently being taken from process.env of your UI)?
Refresh your page and try again, or try removing your plugin configuration and reconfiguring."
`;
